test_fms/diag_manager/test_diag_object_container.F90

```diff
@@ -0,0 +1,234 @@
+      !! vname =  pobj%get_varname()  !! print ...
+      sum = sum + id
+      ic_status = iter%next()
+  end do
+
+  if( sum  /=  full_id_sum) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Container incorrect id sums via iteration")
+  endif
+
+  if( container%size() /= num_objs) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Container incorrect size after inserts")
+  endif
+
+
+  !! Test a removal ****
+  iter = container%iterator()
+  iter  = container%remove( 4, iter )
+  iter  = container%find(4)
+  !! Verify  the removal , part 1:
+  if (  iter%has_data() .eqv. .true.) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Found object of id = 4 after removing it")
+  endif
+   !! Verify  the removal , part 2 :
+  if (container%size() /= (num_objs - 1)) then
+     test_passed = .false.
+    call mpp_error(FATAL,"the_container%size() \= num_obj -1 after a removal ")
+  endif
+
+   !! Verify  the removal , part 3 :
+   !! Iterate over all the objects in the container AFTER the removal of id=4 object;
+  sum = 0
+  iter = container%iterator()
+  do while( iter%has_data() .eqv. .true.)
+      pobj => iter%get()  !!Note use of pointer and pointer assignment is preferred.
+      id =  pobj%get_id( )
+      !! vname =  pobj%get_varname()  !! print ...
+      sum = sum + id
+      ic_status = iter%next()
+  end do
+  if( sum  /=  full_id_sum - 4) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Container incorrect id sums post removal of 4")
+  endif
+  !! End test a removal ****
+
+  !! Test find and access object in the container
+  iter = container%find(7)
+  if (iter%has_data() .neqv. .true. ) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Container did not find obj of id=7")
+  endif
+  !! Checkk the find results more :
+  pobj => iter%get()
+  if(pobj%get_id() /=  7) then
+    test_passed = .false.
+    call mpp_error(FATAL," Id of returned obejct was not 7 ")
+  endif
+  !!TODO further access tests.
+
+
+  !! Manually clear out the container.
+  !! NOTE: In normal use this is NOT PERFORMED since with its finalize function, the  container
+  !! deallocates all pointers and data it manages. However, the client needs to take care of
+  !! the diag objects the client has decided that the container should not manage.
+  !! In this wierd test case, all the diag objects were originally from a vector (a container itself!)
+  !! and not allocated on the heap one at a time, so this step is needed before program completion.
+  do id = 1, num_objs
+    iter  = container%find(id)
+    if (  iter%has_data() .eqv. .true.) then
+      iter  = container%remove( id, iter )
+    endif
+  end do
+
+  if( container%size() /= 0) then
+    test_passed = .false.
+    call mpp_error(FATAL, "Container incorrect size after clearing")
+  endif
+
+  write (6,*) "Finishing diag_obj_container test."
+
+  !! the container has a finalize/destructor which will
+deallocate(container)
+
+call MPI_finalize(ierr)
```

Comment:
I looked around and most mpp test routines from the last two years used MPI_finalize instead of mpp_exit. Is is, it works wiithout printing needless info.